### PR TITLE
feat: add safe upgrade check version 1

### PR DESCRIPTION
**Pull request by jaschdoc**
Adds the `isSafe` predicate.

Related to https://github.com/flix/flix/issues/9525
Related to https://github.com/flix/flix/issues/11436
Related to https://github.com/flix/flix/issues/9524
Related to https://github.com/flix/flix/pull/9730

### CODE REVIEW

**Comment by jaschdoc** [comment]
@magnus-madsen tentative review? I still need add tests and such but so far this is what I have adapted from the old implementation.

EDIT: I just want to make sure that it seem sensible so far.

**Comment by jaschdoc** [comment]
RFR (modulo the debug statements)

**Comment by magnus-madsen** [comment]
@AStenbaek Can you do an initial review? Thanks

**Comment by AStenbaek** [comment]
Looks alright to me. I believe the code represents what we want to check, and the tests cover what I would expect. 
A minor note, I would really like for the test names to actually describe what is being tested — which property? I am mostly concerned that if anyone has to change this in the future they have to spend extra time discerning what is failing.

**Comment by jaschdoc** [comment]
I have made the `isSubset` predicate recursively defined which holds in the case for concrete types but not for polymorphic return types. Intuitively it should hold that if "you can do less than before" then it is okay.
I will use the names `original` and `upgrade` for types that first have type `original` and is then upgraded to `upgrade`.

Consider the types
```flix
original : (Int32 -> Int32 \ {A, B, C}) -> Int32 -> Int32 \ {A, B, C}
upgrade  : (Int32 -> Int32 \ {A, B}) -> Int32 -> Int32 \ {A, B, C}
```
This upgrade should be fine since we can construct the same transcript regardless of the upgraded type:
```flix
def f(g: Int32 -> Int32 \ {A, B}, x: Int32) Int32 \ {A, B, C} = 
    if (false)
        g(x)
    else
        // do A, B, C effects
        0
```
Note that the definition of `f` will result in the same transcript even if `g` has effects `{A, B, C}`.

Now consider the types
```flix
original : (a -> Int32 \ {A, B, C}) -> a -> Int32 \ {A, B, C}
upgrade  : (a -> Int32 \ {A, B}) -> a -> Int32 \ {A, B, C}
```
We can do the same trick again:
```flix
def f(g: a -> Int32 \ {A, B}, x: a) Int32 \ {A, B, C} = 
    if (false)
        g(x)
    else
        // do A, B, C effects
        0
```

But now consider the types 
```flix
original : (a -> b \ {A, B, C}) -> a -> b \ {A, B, C}
upgrade  : (a -> b \ {A, B}) -> a -> b \ {A, B, C}
```
There is no way to construct an element of type `b` unless we have the witness `(a -> b \ {A, B})`. Yes, we can have the effect set `{A, B, C}` in the return type but not without using `g` which changes the transcript after the upgrade since it does strictly less than before. However, my "hope" is that this last case should be okay to upgrade to even though we cannot construct a proof. Why? In the `original` version, it could perform any of `{A, B, C}`, however, it wanted in addition to the `{A, B, C}` inherited by the lambda. In the `upgrade` version it still performs the `{A, B, C}` as before but only inherits `{A, B}` which is still strictly less than before. This is what the recursive definition allows. Note that we can define the function as
```flix
def f(g: a -> b \ {A, B}, x: a): b \ {A, B, C} =
    // do A, B, C effects
    g(x)
```

Lastly, how does this interact with effect variables?
```flix
original : (a -> b \ {A, B, C, ef}) -> a -> b \ {A, B, C, ef}
upgrade  : (a -> b \ {A, B, ef}) -> a -> b \ {A, B, C, ef}
```
This should be fine by the same token as above since it just means that the lambda is invoked (not necessarily on a reachable path) but there is no way to obtain `ef` without doing so. Additionally, it is still free to perform `{A, B, C}` with and without the lambda.

```flix
original : (a -> b \ {A, B, C, ef}) -> a -> b \ {A, B, C, ef}
upgrade  : (a -> b \ {A, B, C}) -> a -> b \ {A, B, C}
```
This does still not allow the function to ignore the lambda since there is no way to construct an element of `b` without it, and it is still free to perform `{A, B, C}` regardless of the lambda.

Lastly, note that the following cannot happen since the types are no longer equal (which would allow it to completely ignore the lambda):
```flix
original : (a -> Int32 \ {A, B, C, ef}) -> a -> Int32 \ {A, B, C, ef}
upgrade  : (a -> Int32 \ {A, B, C}) -> a -> Int32 \ {A, B, C}
```

Maybe for now the correct approach is to use a definition that 100% satisfies the formal definition (i.e., no recursive relation) and then later we can see if can add more flexibility by extending it this way.

**Comment by magnus-madsen** [comment]
@jaschdoc @AStenbaek Does this implement exactly what is in the paper? No more, and no less?

File: main/src/ca/uwaterloo/flix/api/effectlock/EffectUpgrade.scala [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+import ca.uwaterloo.flix.language.phase.unification.{EffUnification3, EqualityEnv}
+import ca.uwaterloo.flix.util.Result
+import ca.uwaterloo.flix.util.collection.ListOps
+
+import scala.collection.immutable.SortedSet
+import scala.collection.mutable
+
+object EffectUpgrade {
+
+  private def debug(obj: Object): Unit = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/effectlock/EffectUpgrade.scala]: Presumably this goes away

File: main/src/ca/uwaterloo/flix/api/effectlock/EffectUpgrade.scala [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  }
+
+  /**
+    * Returns `true` if `upgrade` is a safe upgrade of `original`.
+    *
+    * `upgrade` is a safe upgrade of `original` if at least one of the following holds:
+    *   - `upgrade` is unifiable with `original`.
+    *   - `upgrade` is a monomorphic downgrade of `original`.
+    */
+  def isSafe(original: Scheme, upgrade: Scheme)(implicit flix: Flix): Boolean = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/effectlock/EffectUpgrade.scala]: -> isEffSafeUpgrade

File: main/test/ca/uwaterloo/flix/api/TestEffectLock.scala [hunk 3]
```diff
@@ -1,12 +1,534 @@
...
 import ca.uwaterloo.flix.TestUtils
+import ca.uwaterloo.flix.api.effectlock.EffectUpgrade
 import ca.uwaterloo.flix.api.effectlock.serialization.*
+import ca.uwaterloo.flix.language.ast.TypedAst
 import ca.uwaterloo.flix.util.Options
 import org.scalatest.funsuite.AnyFunSuite
 
 class TestEffectLock extends AnyFunSuite with TestUtils {
 
+  test("Safe.00") {
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/api/TestEffectLock.scala]: Safe.00 -> IsEffectSafeUpgrade.01

File: main/src/ca/uwaterloo/flix/api/effectlock/EffectUpgrade.scala [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    *
+    * 𝜑 ∪ 𝜑' ≡ 𝜑
+    * ----------
+    * 𝜏1 −→ 𝜏2 \ 𝜑 ⪯ 𝜏1 -→ 𝜏2 \ 𝜑′
+    *
+    * `𝜑` are the effects of `original` and `𝜑'` are the effects of `upgrade`.
+    *
+    * Assumes that `original` and `upgrade` have been alpha-renamed so the variables have the same names if they are equal.
+    */
+  private def isSubset(original: Scheme, upgrade: Scheme)(implicit flix: Flix): Boolean = {
```
Comment (AStenbaek) [inline main/src/ca/uwaterloo/flix/api/effectlock/EffectUpgrade.scala:73]: `isSubset` seems to return `true` for something like `isSubset(Int32, Int32)`. Given the rule presumes both types are function types, I believe this should be `false`. We have the generalize rule to handle those cases.

File: main/src/ca/uwaterloo/flix/api/effectlock/EffectUpgrade.scala [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+    case (_, _) =>
+      // Base case: Non-arrow types. Directly compare for equality
+      debug(s"Non-arrow type: $original")
+      debug(s"Non-arrow type: $upgrade")
+      original.base == upgrade.base
+  }
+
+  /**
+    * Checks that the effects of `upgrade` is a subset of the effects of `original`.
```
Comment (AStenbaek) [inline main/src/ca/uwaterloo/flix/api/effectlock/EffectUpgrade.scala:115]: I suggest noting that we check a subset by checking for 𝜑 ∪ 𝜑' ≡ 𝜑, so anyone looking at thin the future knows what is going on.

**Comment by magnus-madsen** [comment]
Can both of you also take a look at `Scheme.lessThanEqual` and tell me if it is relevant/related.

**Comment by jaschdoc** [comment]
> Can both of you also take a look at `Scheme.lessThanEqual` and tell me if it is relevant/related.

It seems relevant. IIRC, we used `Scheme.equal` in the last prototype, so indirectly we also used `Scheme.lessThanEqual`. Why are we not able to use `lessThanEqual` for the upgrade relation (i.e., the upgrade must be less than or equal to the old version)? @AStenbaek 

EDIT: I'm not actually sure we can use `Scheme.lessThanEqual` since we want equality for all types except the effect sets (we also just used `==` to check equality of types, not `Scheme.lessThanOrEqual`).

**Comment by jaschdoc** [comment]
The current version implements exactly the relation in the paper. It seems, however, that we *should* be able to use `Scheme.leq` for the effect subset check. That would hopefully also allow parameters to decrease in effects.

**Comment by jaschdoc** [comment]
Should we add the ignored tests as negative tests? This version intentionally does not handle changes in the parameters (in the subset relation).

File: main/src/ca/uwaterloo/flix/api/effectlock/Util.scala [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import scala.collection.mutable
+
+object Util {
+
+  /**
+    * Performs alpha-renaming on `sc0`.
+    *
+    * To account for change in signatures, symbols with different kinds are renamed differently.
+    */
+  def alpha(sc0: Scheme): Scheme = {
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/api/effectlock/Util.scala:14]: We can maybe move this function to `Scheme`?

**Comment by magnus-madsen** [comment]
@AStenbaek can you review and "sign-off" on correctness? :)

File: main/src/ca/uwaterloo/flix/api/effectlock/Util.scala (hunk 6 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/effectlock/Util.scala:14]: I would keep it here. Remember some of this code goes away again.

**Comment by magnus-madsen** [comment]
> Should we add the ignored tests as negative tests? This version intentionally does not handle changes in the parameters (in the subset relation).

What do you mean? Sorry I don't understand.

**Comment by magnus-madsen** [comment]
@AStenbaek Also see if you can answer the question about Scheme.isLessEqual.

**Comment by magnus-madsen** [comment]
When @AStenbaek signs off, you can merge.

**Review by magnus-madsen — APPROVED** [review]

**Comment by AStenbaek** [comment]
In relation to `Scheme.leq`. It looks to me like this is the relation we want in the generalize rule. I am not convinced this will work for subset, at least when we have base-effects, since it looks to me like this function is about instantiating variables, but I may be wrong.
